File: frontend/Types/Node.ts
import { Comment } from './Comment'
import { Job } from './Job'
import { PageInfo } from './PageInfo'
import { Poll } from './Poll'
import { PollOption } from './PollOption'
import { Story } from './Story'

export type Item = Story | Job | Comment | Poll | PollOption
export type NodeT<T> = T
export interface EdgeT<T> {
  node: NodeT<T>
}
export interface ConnectionT<T> {
  edges: EdgeT<T>[]
  pageInfo: PageInfo
}

File: backend/graphql/node.go
package graphql

import (
	"fmt"

	"github.com/pkg/errors"
	"github.com/tasylab/hn.hsblhsn.me/backend/graphql/internal/hackernews"
	"github.com/tasylab/hn.hsblhsn.me/backend/graphql/internal/msgerr"
	"github.com/tasylab/hn.hsblhsn.me/backend/graphql/model"
)

func ItemToNode(resp *hackernews.ItemResponse) (model.Node, error) {
	switch resp.Type {
	case hackernews.ItemTypeStory:
		return &model.Story{ItemResponse: resp}, nil
	case hackernews.ItemTypeComment:
		return &model.Comment{ItemResponse: resp}, nil
	case hackernews.ItemTypeJob:
		return &model.Job{ItemResponse: resp}, nil
	case hackernews.ItemTypePoll:
		return &model.Poll{ItemResponse: resp}, nil
	case hackernews.ItemTypePollOption:
		return &model.PollOption{ItemResponse: resp}, nil
	default:
		msg := fmt.Sprintf("unknown item type: %q", resp.Type)
		return nil, msgerr.New(errors.New(msg), "Invalid item type received")
	}
}
